usm_wrappers/utils.py

```diff
@@ -0,0 +1,513 @@
+                return False
+            self._processStreams()
+        return True
+
+    def communicate(self, data=None):
+        if data is not None:
+            self.stdin.write(data)
+            self.stdin.flush()
+        self.stdin.close()
+
+        self.wait()
+        return "".join(self.stdout), "".join(self.stderr)
+
+    def __del__(self):
+        self._poller.close()
+
+
+class CmdExecFailed(Exception):
+    message = "command execution failed"
+
+    def __init__(self, cmd, rc, out=(), err=()):
+        self.cmd = cmd
+        self.rc = rc
+        self.out = out
+        self.err = err
+
+    def __str__(self):
+        s = "%s\ncommand: %s\nexit code: %s\nstderr: %s\nstdout: %s\n"
+        return s % (self.message, self.cmd, self.rc, self.err, self.out)
+
+
+def execCmd(command, sudo=False, cwd=None, data=None, raw=True,
+            printable=None, env=None, sync=True, nice=None, ioclass=None,
+            ioclassdata=None, setsid=False, execCmdLogger=logging.root,
+            deathSignal=0, childUmask=None, throwException=True):
+    """
+    Executes an external command, optionally via sudo.
+
+    IMPORTANT NOTE: the new process would receive `deathSignal` when the
+    controlling thread dies, which may not be what you intended: if you create
+    a temporary thread, spawn a sync=False sub-process, and have the thread
+    finish, the new subprocess would die immediately.
+    """
+    if ioclass is not None:
+        cmd = command
+        command = [ioniceCmdPath.cmd, '-c', str(ioclass)]
+        if ioclassdata is not None:
+            command.extend(("-n", str(ioclassdata)))
+
+        command = command + cmd
+
+    if nice is not None:
+        command = [niceCmdPath.cmd, '-n', str(nice)] + command
+
+    if setsid:
+        command = [setsidCmdPath.cmd] + command
+
+    if sudo:
+        command = [sudoCmdPath.cmd, SUDO_NON_INTERACTIVE_FLAG] + command
+
+    if not printable:
+        printable = command
+
+    cmdline = repr(subprocess.list2cmdline(printable))
+    execCmdLogger.debug("%s (cwd %s)", cmdline, cwd)
+
+    p = CPopen(command, close_fds=True, cwd=cwd, env=env,
+               deathSignal=deathSignal, childUmask=childUmask)
+    p = AsyncProc(p)
+    if not sync:
+        if data is not None:
+            p.stdin.write(data)
+            p.stdin.flush()
+
+        return p
+
+    (out, err) = p.communicate(data)
+
+    if out is None:
+        # Prevent splitlines() from barfing later on
+        out = ""
+
+    execCmdLogger.debug("%s: <err> = %s; <rc> = %d",
+                        {True: "SUCCESS", False: "FAILED"}[p.returncode == 0],
+                        repr(err), p.returncode)
+
+    if throwException and p.returncode:
+        raise CmdExecFailed(command, p.returncode, out, err)
+
+    if not raw:
+        out = out.splitlines(False)
+        err = err.splitlines(False)
+
+    return (p.returncode, out, err)
+
+
+paramiko.util.get_logger('paramiko').setLevel(logging.ERROR)
```

Comment:
Basically we would need to set log level of paramiko.  I felt the util module should take care of it as api consumer doesn't care about paramiko whereas in rexecCmd().  Do you think this has to be in consumer level?